File: frameworks/go-graphql-go/error_scenarios_test.go
package main

import (
	"testing"
)

// ============================================================================
// Error: Not Found
// ============================================================================

func TestNotFoundErrors(t *testing.T) {
	tests := []struct {
		name      string
		testType  string
	}{
		{
			name:      "user not found returns nil",
			testType:  "user",
		},
		{
			name:      "post not found returns nil",
			testType:  "post",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			// Act - try to get nonexistent entity
			if tt.testType == "user" {
				user := factory.GetUser("nonexistent-id")
				if user != nil {
					t.Errorf("Expected nil for nonexistent user")
				}
			} else if tt.testType == "post" {
				post := factory.GetPost("nonexistent-id")
				if post != nil {
					t.Errorf("Expected nil for nonexistent post")
				}
			}
		})
	}
}

// ============================================================================
// Error: Invalid Input
// ============================================================================

func TestInvalidInputHandling(t *testing.T) {
	tests := []struct {
		name      string
		limit     int
		shouldClamp bool
	}{
		{
			name:       "limit 0 handled",
			limit:      0,
			shouldClamp: true,
		},
		{
			name:       "negative limit clamped",
			limit:      -5,
			shouldClamp: true,
		},
		{
			name:       "very large limit capped",
			limit:      999999,
			shouldClamp: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange & Act
			result := tt.limit
			if result < 0 {
				result = 0
			}
			if result > 100 {
				result = 100
			}

			// Assert
			if tt.shouldClamp && result != tt.limit {
				if tt.limit < 0 && result != 0 {
					t.Errorf("Negative limit should be clamped to 0")
				}
				if tt.limit > 100 && result != 100 {
					t.Errorf("Large limit should be capped at 100")
				}
			}
		})
	}
}

// ============================================================================
// Error: Data Type Validation
// ============================================================================

func TestDataTypeValidation(t *testing.T) {
	tests := []struct {
		name      string
		testCase  string
		isValid   bool
	}{
		{
			name:     "valid UUID format",
			testCase: "uuid",
			isValid:  true,
		},
		{
			name:     "username is string type",
			testCase: "string",
			isValid:  true,
		},
		{
			name:     "empty username invalid",
			testCase: "empty",
			isValid:  false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			validator := NewValidationHelper(t)

			// Act & Assert
			if tt.testCase == "uuid" {
				testID := "123e4567-e89b-12d3-a456-426614174000"
				validator.AssertUUID(testID)
			} else if tt.testCase == "string" {
				username := "alice"
				if username == "" {
					t.Errorf("Username should not be empty")
				}
			} else if tt.testCase == "empty" {
				username := ""
				if username != "" {
					t.Errorf("Username is empty")
				}
			}
		})
	}
}

// ============================================================================
// Error: Null Field Consistency
// ============================================================================

func TestNullFieldConsistency(t *testing.T) {
	tests := []struct {
		name         string
		bio          string
		wantNil      bool
	}{
		{
			name:    "null bio consistency",
			bio:     "",
			wantNil: true,
		},
		{
			name:    "present bio",
			bio:     "My bio",
			wantNil: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			user := factory.CreateTestUser("alice", "alice@example.com", "Alice", tt.bio)

			// Assert
			if tt.wantNil {
				if user.Bio != nil {
					t.Errorf("Expected nil bio")
				}
			} else {
				if user.Bio == nil {
					t.Errorf("Expected non-nil bio")
				}
			}
		})
	}
}

// ============================================================================
// Error: Special Character Handling
// ============================================================================

func TestSpecialCharacterErrors(t *testing.T) {
	tests := []struct {
		name    string
		content string
	}{
		{
			name:    "single quotes",
			content: "I'm a developer",
		},
		{
			name:    "double quotes",
			content: `He said "hello"`,
		},
		{
			name:    "HTML tags",
			content: "Check <this> out",
		},
		{
			name:    "ampersand",
			content: "Tom & Jerry",
		},
		{
			name:    "emoji",
			content: "🎉 Celebration! 🚀 Rocket",
		},
		{
			name:    "accents",
			content: "Àlice Müller",
		},
		{
			name:    "diacritics",
			content: "José García",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			// Act
			post := factory.CreateTestPost("user-id", "Title", tt.content)

			// Assert
			if post == nil {
				t.Errorf("Post creation failed")
			}
		})
	}
}

// ============================================================================
// Error: Boundary Conditions
// ============================================================================

func TestBoundaryConditionErrors(t *testing.T) {
	tests := []struct {
		name    string
		length  int
		isValid bool
	}{
		{
			name:    "very long bio (5000 chars)",
			length:  5000,
			isValid: true,
		},
		{
			name:    "very long username (255 chars)",
			length:  255,
			isValid: true,
		},
		{
			name:    "very long post content (5000 chars)",
			length:  5000,
			isValid: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			generator := NewDataGenerator(t)

			// Act
			content := generator.GenerateLongString(tt.length)

			// Assert
			if tt.isValid {
				if len(content) != tt.length {
					t.Errorf("Expected %d chars, got %d", tt.length, len(content))
				}
			}
		})
	}
}

// ============================================================================
// Error: Response Structure
// ============================================================================

func TestResponseStructureValidation(t *testing.T) {
	tests := []struct {
		name       string
		entityType string
	}{
		{
			name:       "user response has required fields",
			entityType: "user",
		},
		{
			name:       "post response has required fields",
			entityType: "post",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			// Act & Assert
			if tt.entityType == "user" {
				user := factory.CreateTestUser("alice", "alice@example.com", "Alice", "")

				if user.ID == "" {
					t.Errorf("User ID missing")
				}
				if user.Username != "alice" {
					t.Errorf("Username missing")
				}
			} else if tt.entityType == "post" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Post", "Content")

				if post.ID == "" {
					t.Errorf("Post ID missing")
				}
				if post.Title != "Post" {
					t.Errorf("Title missing")
				}
			}
		})
	}
}

// ============================================================================
// Error: Uniqueness Validation
// ============================================================================

func TestUniquenessValidation(t *testing.T) {
	tests := []struct {
		name string
	}{
		{
			name: "multiple users have unique IDs",
		},
		{
			name: "multiple posts have unique IDs",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			if tt.name == "multiple users have unique IDs" {
				user1 := factory.CreateTestUser("alice", "alice@example.com", "Alice", "")
				user2 := factory.CreateTestUser("bob", "bob@example.com", "Bob", "")
				user3 := factory.CreateTestUser("charlie", "charlie@example.com", "Charlie", "")

				// Assert
				if user1.ID == user2.ID || user2.ID == user3.ID || user1.ID == user3.ID {
					t.Errorf("User IDs are not unique")
				}
			} else if tt.name == "multiple posts have unique IDs" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post1 := factory.CreateTestPost(author.ID, "Post1", "Content1")
				post2 := factory.CreateTestPost(author.ID, "Post2", "Content2")

				if post1.ID == post2.ID {
					t.Errorf("Post IDs are not unique")
				}
			}
		})
	}
}

// ============================================================================
// Error: Relationship Integrity
// ============================================================================

func TestRelationshipIntegrity(t *testing.T) {
	tests := []struct {
		name string
	}{
		{
			name: "post author ID is valid UUID",
		},
		{
			name: "post references correct author",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			validator := NewValidationHelper(t)
			defer factory.Reset()

			if tt.name == "post author ID is valid UUID" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Post", "Content")

				if post.AuthorID == nil {
					t.Errorf("AuthorID is nil")
					return
				}
				validator.AssertUUID(*post.AuthorID)
			} else if tt.name == "post references correct author" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Post", "Content")

				if post.AuthorID == nil || *post.AuthorID != author.ID {
					t.Errorf("Post does not reference correct author")
				}
			}
		})
	}
}

// ============================================================================
// Error: Data Consistency Across Requests
// ============================================================================

func TestDataConsistencyErrors(t *testing.T) {
	tests := []struct {
		name string
	}{
		{
			name: "same user returns same data",
		},
		{
			name: "same post returns same data",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			if tt.name == "same user returns same data" {
				user := factory.CreateTestUser("alice", "alice@example.com", "Alice", "Developer")

				// Act
				retrieved1 := factory.GetUser(user.ID)
				retrieved2 := factory.GetUser(user.ID)

				// Assert
				if retrieved1 == nil || retrieved2 == nil {
					t.Errorf("User not found")
					return
				}
				if retrieved1.ID != retrieved2.ID {
					t.Errorf("User ID changed")
				}
				if retrieved1.Username != retrieved2.Username {
					t.Errorf("Username changed")
				}
			} else if tt.name == "same post returns same data" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Test", "Content")

				// Act
				retrieved1 := factory.GetPost(post.ID)
				retrieved2 := factory.GetPost(post.ID)

				// Assert
				if retrieved1 == nil || retrieved2 == nil {
					t.Errorf("Post not found")
					return
				}
				if retrieved1.Title != retrieved2.Title {
					t.Errorf("Post title changed")
				}
			}
		})
	}
}

File: frameworks/async-graphql/tests/error_scenarios_test.rs
mod test_helpers;
use test_helpers::*;

// ============================================================================
// Error: Not Found
// ============================================================================

#[test]
fn test_user_not_found_returns_none() {
    let factory = TestFactory::new();
    let user = factory.get_user("nonexistent-id");
    assert!(user.is_none());
}

#[test]
fn test_post_not_found_returns_none() {
    let factory = TestFactory::new();
    let post = factory.get_post("nonexistent-id");
    assert!(post.is_none());
}

// ============================================================================
// Error: Invalid Input
// ============================================================================

#[test]
fn test_limit_zero_handled() {
    let limit = 0;
    let clamped = if limit < 0 { 0 } else { limit };
    assert_eq!(clamped, 0);
}

#[test]
fn test_negative_limit_clamped() {
    let limit = -5;
    let clamped = if limit < 0 { 0 } else { limit };
    assert_eq!(clamped, 0);
}

#[test]
fn test_very_large_limit_capped() {
    let limit = 999999;
    let clamped = if limit > 100 { 100 } else { limit };
    assert_eq!(clamped, 100);
}

// ============================================================================
// Error: Data Type Validation
// ============================================================================

#[test]
fn test_valid_uuid_format() {
    let test_id = "123e4567-e89b-12d3-a456-426614174000";
    assert!(ValidationHelper::assert_uuid(test_id).is_ok());
}

#[test]
fn test_username_is_string_type() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    assert!(!user.username.is_empty());
}

#[test]
fn test_empty_username_invalid() {
    let username = "";
    assert!(username.is_empty());
}

// ============================================================================
// Error: Null Field Consistency
// ============================================================================

#[test]
fn test_null_bio_consistency() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    assert_eq!(user.bio, None);
}

#[test]
fn test_present_bio() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "My bio");
    assert_eq!(user.bio, Some("My bio".to_string()));
}

// ============================================================================
// Error: Special Character Handling
// ============================================================================

#[test]
fn test_single_quotes() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "I'm a developer");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_double_quotes() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "He said \"hello\"");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_html_tags() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "Check <this> out");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_ampersand() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "Tom & Jerry");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_emoji() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "🎉 Celebration! 🚀 Rocket");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_accents() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Àlice Müller", "");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_diacritics() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "José García", "");
    assert!(factory.get_user(&user.id).is_some());
}

// ============================================================================
// Error: Boundary Conditions
// ============================================================================

#[test]
fn test_very_long_bio_5000_chars() {
    let factory = TestFactory::new();
    let long_bio = DataGenerator::generate_long_string(5000);
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", &long_bio);
    let retrieved = factory.get_user(&user.id).unwrap();
    assert_eq!(retrieved.bio.as_ref().unwrap().len(), 5000);
}

#[test]
fn test_very_long_username_255_chars() {
    let factory = TestFactory::new();
    let long_name = DataGenerator::generate_long_string(255);
    let user = factory.create_test_user(&long_name, "user@example.com", "User", "");
    let retrieved = factory.get_user(&user.id).unwrap();
    assert_eq!(retrieved.username.len(), 255);
}

#[test]
fn test_very_long_post_content_5000_chars() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let long_content = DataGenerator::generate_long_string(5000);
    let post = factory.create_test_post(&author.id, "Title", &long_content);
    let retrieved = factory.get_post(&post.id).unwrap();
    assert_eq!(retrieved.content.len(), 5000);
}

// ============================================================================
// Error: Response Structure
// ============================================================================

#[test]
fn test_user_response_has_required_fields() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");

    assert!(!user.id.is_empty());
    assert_eq!(user.username, "alice");
}

#[test]
fn test_post_response_has_required_fields() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");

    assert!(!post.id.is_empty());
    assert_eq!(post.title, "Post");
}

// ============================================================================
// Error: Uniqueness Validation
// ============================================================================

#[test]
fn test_multiple_users_have_unique_ids() {
    let factory = TestFactory::new();
    let user1 = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let user2 = factory.create_test_user("bob", "bob@example.com", "Bob", "");
    let user3 = factory.create_test_user("charlie", "charlie@example.com", "Charlie", "");

    assert_ne!(user1.id, user2.id);
    assert_ne!(user2.id, user3.id);
    assert_ne!(user1.id, user3.id);
}

#[test]
fn test_multiple_posts_have_unique_ids() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post1 = factory.create_test_post(&author.id, "Post1", "Content1");
    let post2 = factory.create_test_post(&author.id, "Post2", "Content2");

    assert_ne!(post1.id, post2.id);
}

// ============================================================================
// Error: Relationship Integrity
// ============================================================================

#[test]
fn test_post_author_id_is_valid_uuid() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");

    assert!(ValidationHelper::assert_uuid(&author.id).is_ok());
}

#[test]
fn test_post_references_correct_author() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");

    assert_eq!(post.fk_author, author.pk_user);
}

// ============================================================================
// Error: Data Consistency Across Requests
// ============================================================================

#[test]
fn test_same_user_returns_same_data() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");

    let retrieved1 = factory.get_user(&user.id);
    let retrieved2 = factory.get_user(&user.id);

    assert_eq!(retrieved1.as_ref().unwrap().id, retrieved2.as_ref().unwrap().id);
    assert_eq!(
        retrieved1.as_ref().unwrap().username,
        retrieved2.as_ref().unwrap().username
    );
}

#[test]
fn test_same_post_returns_same_data() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Test", "Content");

    let retrieved1 = factory.get_post(&post.id);
    let retrieved2 = factory.get_post(&post.id);

    assert_eq!(retrieved1.as_ref().unwrap().title, retrieved2.as_ref().unwrap().title);
}
